sim/simulation_scenario.py

```diff
@@ -0,0 +1,139 @@
+# pylint: disable=line-too-long, too-many-arguments, too-many-locals
+
+""" Defines classes that allow a scripted simulation scenario to be defined."""
+
+from abc import ABC, abstractmethod
+from queue import SimpleQueue
+from typing import Any, Tuple
+import importlib
+import logging
+import ruamel.yaml
+
+from sim.mock_pin import IMockPin
+from sim.data_provider import IDataProvider
+from sim.mock_device_registry import MockDeviceRegistry
+
+
+class SimulationAction(ABC):
+
+    """ Base class for all action types used with SimulationScenario """
+
+    def __init__(self, name: str):
+        self._name = name
+
+    @abstractmethod
+    def run(self) -> None:
+        """Performs the action"""
+        raise NotImplementedError
+
+
+class SetDataProviderAction(SimulationAction):
+
+    """A simulation action that changes the data_provider an Mock Device Pin"""
+
+    def __init__(self, name: str, pin: IMockPin, data_provider: IDataProvider):
+        super().__init__(name)
+
+        if not data_provider:
+            raise ValueError("A valid data provider must be specified")
+
+        if not pin:
+            raise ValueError(
+                "A valid data provider container must be specified"
+            )
+
+        if not hasattr(pin, "set_data_provider"):
+            raise RuntimeError(
+                "The specified pin is not a valid data_provider container"
+            )
+
+        self._data_provider = data_provider
+        self._data_provider_container = pin
+
+    def run(self) -> None:
+        logging.info("Running %s : %s", self.__class__.__name__, self._name)
+        self._data_provider_container.set_data_provider(self._data_provider)
+
+
+class ActionGroup(SimulationAction):
+
+    """Represents a group of actions that will be performed when this action is run.
+    For example, if you wanted to change the data provider on multiple output pins at once"""
+
+    def __init__(self, name, actions):
+        super().__init__(name)
+        self._actions = list(actions)
+
+    def add_action(self, action: SimulationAction) -> None:
+
+        """Adds an action to this step that will set the specified data provider onto the specified container"""
+        self._actions.append(action)
+
+    def run(self) -> None:
+        logging.info("Running %s : %s", self.__class__.__name__, self._name)
+        for action in self._actions:
+            action.run()
+
+
+class SimulationScenario(SimpleQueue[SimulationAction]):
```

Comment:
This is a bit strange but I think this needs to be types as a string as "SimpleQueue" is not subscribable at runtime in all the python versions we support 

```suggestion
class SimulationScenario("SimpleQueue[SimulationAction]"):
```

Reply:
I ended up encapsulating the SimpleQueue inside SimulationScenario, instead of deriving from it, which allowed me to use the quoted type.